src/main/scala/io/iohk/ethereum/domain/Transaction.scala

```diff
@@ -22,15 +22,31 @@ sealed trait Transaction {
 }
 
 object Transaction {
-  val Type01: Int = 1
+  val Type01: Byte = 1.toByte
+
+  val MinAllowedType: Byte = 0
+  val MaxAllowedType: Byte = 0x7f
+
   val LegacyThresholdLowerBound: Int = 0xc0
   val LegacyThresholdUpperBound: Int = 0xfe
+
+  def withGasLimit(gl: BigInt): Transaction => Transaction = {
+    case tx: LegacyTransaction         => tx.copy(gasLimit = gl)
+    case tx: TransactionWithAccessList => tx.copy(gasLimit = gl)
+  }
+
+  implicit class TransactionTypeValidator(val transactionType: Byte) extends AnyVal {
+    def isValidTransactionType: Boolean = transactionType >= MinAllowedType && transactionType <= MaxAllowedType
```

Comment:
I'm not really a big fan of implicit conversion if we can avoid it because it can be surprising where it is used. Also, currently only `Type01` is allowed. Wouldn't it be more safe to just allow that for now ? 

Or maybe we can move that to the encoding/networking part because the fact that a transaction type is encoded as a number from 0 to 0x7f is not really relevant in the domain. We can have transaction types as an enumeration to make it impossible to have an invalid transaction type in the domain part, and let the networking part make the conversion.

Reply:
I agree with the TransactionType enum:
- only check when decoding RLP data
- easier validation (transactionTypeByte is one of enum values)

As for the validation not being currently restricted to type01, I got two choices:
- current implementation: can accept un-handled transaction type, but don't need to change for new transaction
- proposed implementation: only valid for supported transaction types, but need to be updated for every new transaction

Your version seems more safe.

As discussed, it will be a good enhancement for the Mamba rebase.